src/status_im/chat/models/images.cljs

```diff
@@ -100,13 +100,15 @@
 
 (re-frame/reg-fx
  ::camera-roll-get-photos
- (fn [num]
+ (fn [[num end-cursor]]
    (permissions/request-permissions
     {:permissions [:read-external-storage]
      :on-allowed  (fn []
-                    (-> (.getPhotos CameraRoll #js {:first num :assetType "Photos" :groupTypes "All"})
-                        (.then #(re-frame/dispatch [:on-camera-roll-get-photos (:edges (types/js->clj %))]))
-                        (.catch #(log/warn "could not get cameraroll photos"))))})))
+                    (-> (if end-cursor
```

Comment:
I usually prefer to see effects that abstract the `on-success` and `on-failure` handlers, so they become more general and are easier to test. This is a known pattern, you can see it, for instance in the http-fx effect https://github.com/day8/re-frame-http-fx/blob/master/src/day8/re_frame/http_fx.cljs#L26

But the really important feedback I have for this effect is that it lacks testing, and since you're touching it, it'd be great to have them.

Reply:
I could be wrong, but I do not think we can use `on-success` and `on-failure` here. This is a JS promise.

`permissions/request-permissions` requires reading from an actual device/simulator.

Reply:
@ilmotta Thank you so much for this detailed explanation! ❤️ Lots of useful information.

I have one question, so here we have `.then` followed by `on-success`, like is there any other option? Can't we just drop out the `on-success` and `on-failure` since we are using `then` and `catch`?

Reply:
@OmarBasem, we can definitely drop the `on-success` and `on-failure` options. It's up to the developer to choose how reusable the effect is. If you want to design an effect that other events can dispatch and react accordingly, then having `on-success`/`on-failure` is the recommended approach in re-frame. But if you just want other events to *fire and forget* (dispatch `::camera-roll/get-photos`) then you can drop those options.

Both alternatives are valid. Without `on-success`/`on-failure` it's also a bit harder to test, but still doable.